### DISCUSSION TITLE
Is there a reason why CanvasModulate doesn't respect Item Cull Mask?

### DISCUSSION THREAD

**Discussion by hunterloftis** [godotengine/godot-proposals#7466] [2D]
There are many examples of people trying to exclude nodes from the effects of CanvasModulate:

- https://www.reddit.com/r/godot/comments/n7x1lt/i_there_any_way_to_make_canvas_modulate_node/
- https://www.reddit.com/r/godot/comments/a08a20/how_to_make_canvasmodulate_not_affect_certain/
- https://www.reddit.com/r/godot/comments/fgxwl7/exclude_nodes_from_canvasmodulate/
- https://www.reddit.com/r/godot/comments/hqyx1n/how_to_exclude_a_sprite_from_canvusmodulate/
- etc

The most-commonly-recommended solution is to put excluded content inside a separate CanvasLayer. This has lots of drawbacks though and is sometimes impossible, since it requires rearranging the node tree. The only alternative I have found is to replace the CanvasItem material with an unshaded one, for every item to be excluded.

Lights have a simple mechanism to allow canvas items to opt-out of being lit: `range_item_cull_mask`. Is there a reason why `CanvasModulate` doesn't use the same mechanism to allow items to opt-out of being modulated?

**Comment by Calinou** [comment]
> Is there a reason why CanvasModulate doesn't use the same mechanism to allow items to opt-out of being modulated?

Internally, this is all CanvasModulate does: https://github.com/godotengine/godot/blob/eca6f0eb545166708c56633152974f40f51de2be/scene/2d/canvas_modulate.cpp#L73

This would require implementing per-layer modulation, which sounds technically feasible, but it may have a performance cost (probably not a huge one though).

**Reply by hunterloftis** [reply]
I see, thanks for the explanation & the link!
